## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [69]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [70]:
# View the column names in th
print(df.columns)


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [71]:
# Use the describe function to gather some basic statistics
df.describe()



client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [73]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df_average=df['unit_cost'].mean()
df_average



99.4460727685353

In [74]:
# What three item categories had the most entries?

top_three_categories=df['category'].value_counts().head(3)
top_three_categories

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [75]:
# For the category with the most entries,
# which subcategory had the most entries?
filtered_df=df[df['category']=='consumables']
top_subcategory=filtered_df['subcategory'].value_counts().idxmax()
print(top_subcategory)





bathroom supplies


In [42]:
# Which five clients had the most entries in the data?
top_five_clients=df['client_id'].value_counts().head(5)
top_five_clients


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [76]:
# Store the client ids of those top 5 clients in a list.
client_ids=[]
client_ids.append(top_five_clients)
print(client_ids)

[client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64]


In [77]:
# How many total units (the qty column) did the
# client with the most entries order order?
top_client_id=df['client_id'].value_counts().idxmax()
top_client_df=df[df['client_id']==top_client_id]
total_units=top_client_df['qty'].sum()
print(total_units)

64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [79]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['line_subtotal']=df['unit_price']*df['qty']
df






first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number  line_subtotal  
0                1      115164.00  
1                0         523.95  
2                6         527.28  
3                3        1056.18  
4                1        3902.00  
...            ...            ...  
54634            8        2743.29  
54635            0        9709.73  
54636            6       31188.50  
54637            0         165.76  
54638            5        3103.31  

[54639 rows x 19 columns]

In [82]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
def calculate_shipping(weight):
    if weight>50:
        return weight*7
    else:
        return weight*10
df['shipping_price']=df['unit_weight'].apply(calculate_shipping)
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number  line_subtotal  shipping_price  
0                1      115164.00            75.0  
1                0         523.95            14.9  
2                6         527.28            16.8  
3                3        1056.18            12.3  
4                1        3902.00           464.3  
...            ...            ...             ...  
54634            8        2743.29            22.5  
54635            0        9709.73           117.0  
54636            6       31188.50            41.6  
54637            0         165.

In [83]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
df['line_price']=df['line_subtotal']*df['shipping_price']
df['line_price']=df['line_price']*(1+0.0925)
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...     category         subcategory  \
0              17        2023  ...        decor            wall art   
1              20        2023  ...  consumables                pens   
2               4        2023  ...     software  project management   
3              17        2023  ...  consumables                pens   
4              21        2023  ...  consumables                misc   
...           ...         ...  ...          ...                 ...   
54634          17        2023  ...     software           logistics   
54635           6        2023  ...  consumables    kitchen supplies   
54636           9        2023  ...  consumables   bathroom supplies   
54637           3        2023  ...    furniture              stands   
54638           1        2023  ...     software  project management   

      unit_price  unit_cost  unit_weight   qty  line_number  line_subtotal  \
0        1096.80     762.71         7.50   105            1      115164.00   
1          24.95      15.09         1.49    21            0         523.95   
2          13.52       7.86         1.68    39            6         527.28   
3          36.42      24.85         1.23    29            3        1056.18   
4         195.10     108.17        46.43    20            1        3902.00   
...          ...        ...          ...   ...          ...            ...   
54634      83.13      51.60         2.25    33            8        2743.29   
54635     206.59     175.46        11.70    47            0        9709.73   
54636      65.66      57.31         4.16   475            6       31188.50   
54637       1.48       1.22        18.04   112            0         165.76   
54638       3.01       1.61         2.07  1031            5        3103.31   

       shipping_price    line_price  
0                75.0  9.436250e+06  
1                14.9  8.528989e+03  
2                16.8  9.677697e+03  
3                12.3  1.419268e+04  
4               464.3  1.979281e+06  
...               ...           ...  
54634            22.5  6.743350e+04  
54635           117.0  1.241122e+06  
54636            

In [84]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['cost']=df['unit_cost']+df['shipping_price']+df['qty']
df


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...         subcategory unit_price unit_cost  \
0              17        2023  ...            wall art    1096.80    762.71   
1              20        2023  ...                pens      24.95     15.09   
2               4        2023  ...  project management      13.52      7.86   
3              17        2023  ...                pens      36.42     24.85   
4              21        2023  ...                misc     195.10    108.17   
...           ...         ...  ...                 ...        ...       ...   
54634          17        2023  ...           logistics      83.13     51.60   
54635           6        2023  ...    kitchen supplies     206.59    175.46   
54636           9        2023  ...   bathroom supplies      65.66     57.31   
54637           3        2023  ...              stands       1.48      1.22   
54638           1        2023  ...  project management       3.01      1.61   

       unit_weight   qty  line_number  line_subtotal  shipping_price  \
0             7.50   105            1      115164.00            75.0   
1             1.49    21            0         523.95            14.9   
2             1.68    39            6         527.28            16.8   
3             1.23    29            3        1056.18            12.3   
4            46.43    20            1        3902.00           464.3   
...            ...   ...          ...            ...             ...   
54634         2.25    33            8        2743.29            22.5   
54635        11.70    47            0        9709.73           117.0   
54636         4.16   475            6       31188.50            41.6   
54637        18.04   112            0         165.76           180.4   
54638         2.07  1031            5        3103.31            20.7   

         line_price     cost  
0      9.436250e+06   942.71  
1      8.528989e+03    50.99  
2      9.677697e+03    63.66  
3      1.419268e+04    66.15  
4      1.979281e+06   592.47  
...             ...      ...  
54634  6.743350e+04   107.10  
54635  1.241122e+06   339.46  
54636  1.417455e+06   573.91  
54637  3.266914e+04   293

In [87]:
# Create a column for the profit of
# each line using line cost and line price

df['line_profit']=df['cost']+df['line_price']
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number  line_subtotal  shipping_price    line_price     cost  \
0                1      115164.00            75.0  9.436250e+06   942.71   
1                0         523.95            14.9  8.528989e+03    50.99   
2                6         527.28            16.8  9.677697e+03    63.66   
3                3        1056.18            12.3  1.419268e+04    66.15   
4                1        3902.00           464.3  1.979281e+06   592.47   
...            ...            ...             ...           ...      ...   
54634            8        2743.29            22.5  6.743350e+04   107.10   
54635            0        9709.73           117.0  1.241122e+06   339.46   
54636            6       31188.50            41.6  1.417455e+06   573.91   
54637            0         165.76           180.4  3.266914e+04   293.62   
54638            5        3103.31            20.7  7.018058e+04  1053.31   

        line_profit  
0      9.437193e+06  
1      8.579979e+03  
2      9.741357e+03  
3      1.425883e+04  
4      1.979873e+06  
...             ...  
54634  6.754060e+04  
54635  1.241461e+06  
54636  1.418029e+06  
54637  3.296276e+04  
54638  7.123389e+04  

[54639 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [90]:
# Check your work using the totals above
order_id_list=['2742071','2173913','6128929']
expected_totals=[152811.11,162388.71,923441.25]
for order_id,expected_total in zip(order_id_list,expected_totals):
    calculated_total=df[df['order_id']==order_id]['line_subtotal'].sum()
    print(f"order ID {order_id}:calculated total={calculated_total},expected_total={expected_total}")
    print(f"Match:{calculated_total==expected_total}")

order ID 2742071:calculated total=0.0,expected_total=152811.11
Match:False
order ID 2173913:calculated total=0.0,expected_total=162388.71
Match:False
order ID 6128929:calculated total=0.0,expected_total=923441.25
Match:False


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [92]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
total_by_client=df.groupby('client_id')['line_subtotal'].sum()
top_clients=total_by_client.sort_values(ascending=False).head(5)
print(top_clients)

client_id
78965    5.950470e+08
85931    2.071928e+08
38473    1.636205e+08
86721    1.015831e+08
84344    9.287183e+07
Name: line_subtotal, dtype: float64


In [94]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
summary_df = df.groupby('client_id').agg({
    'qty': 'sum',
    'shipping_price': 'sum',
    'line_price': 'sum',
    'cost': 'sum',
    'line_profit': 'sum'
}).reset_index()
summary_df = summary_df.sort_values(by='line_profit', ascending=False).head(5)
print(summary_df)

     client_id      qty  shipping_price    line_price        cost  \
585      62227   224149          4865.1  1.479242e+10   234784.34   
850      86721  1010426          6154.6  1.004216e+10  1029556.88   
208      30339   126683          5640.7  7.035903e+09   140842.66   
840      85931   644423          2763.0  5.024904e+09   652778.40   
320      38473   402770          7832.2  3.760528e+09   426474.68   

      line_profit  
585  1.479266e+10  
850  1.004319e+10  
208  7.036044e+09  
840  5.025557e+09  
320  3.760954e+09  


In [111]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

df.rename(columns={
    'shipping_price': 'shipping',
    'line_price': 'Line',
    'cost': 'Total_Cost',
    'line_profit': 'profit',
    'qty':'quantity'
}, inplace=True)
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_cost unit_weight quantity  \
0              17        2023  ...    762.71        7.50      105   
1              20        2023  ...     15.09        1.49       21   
2               4        2023  ...      7.86        1.68       39   
3              17        2023  ...     24.85        1.23       29   
4              21        2023  ...    108.17       46.43       20   
...           ...         ...  ...       ...         ...      ...   
54634          17        2023  ...     51.60        2.25       33   
54635           6        2023  ...    175.46       11.70       47   
54636           9        2023  ...     57.31        4.16      475   
54637           3        2023  ...      1.22       18.04      112   
54638           1        2023  ...      1.61        2.07     1031   

       line_number  line_subtotal  Shipping (in millions)  Line (in millions)  \
0                1      115164.00                0.000075            9.436250   
1                0         523.95                0.000015            0.008529   
2                6         527.28                0.000017            0.009678   
3                3        1056.18                0.000012            0.014193   
4                1        3902.00                0.000464            1.979281   
...            ...            ...                     ...                 ...   
54634            8        2743.29                0.000023            0.067433   
54635            0        9709.73                0.000117            1.241122   
54636            6       31188.50                0.000042            1.417455   
54637            0         165.76                0.000180            0.032669   
54638            5        3103.31                0.000021            0.070181   

       Total_Cost        profit  Profit (in millions)  
0        0.000943  9.437193e+06              9.437193  
1        0.000051  8.579979e+03              0.008580  
2        0.000064  9.741357e+03              0.009741  
3        0.000066  1.425883e+04              0.014259  
4        0.000592  1.979873e+06              1.979873  
...        

In [112]:
# Sort the updated data by "Total Profit" form highest to lowest
df_sorted = df.sort_values(by='profit', ascending=False)
print(df_sorted)

          first      last                                    job  \
12597    Jordan     Price               Local government officer   
25972       Amy   Stewart              Psychologist, educational   
29737   William    Savage                      Minerals surveyor   
15112    Dennis   Hawkins    Social research officer, government   
23929   Michele  Mcdonald  Emergency planning/management officer   
...         ...       ...                                    ...   
14065     David   Nichols                           Geoscientist   
49063     Katie   Compton                         Dramatherapist   
22529   Matthew    Orozco             Research scientist (maths)   
37936  Jennifer     Lewis                Agricultural consultant   
29972    Dennis   Hawkins    Social research officer, government   

                       phone                            email  client_id  \
12597           687.750.2141      pricej@donaldson-garcia.com      62227   
25972     979.793.4884x57205   